datatypes/src/raster/grid_dimension.rs

```diff
@@ -123,301 +201,687 @@ impl<D> Capacity for D
+        self.dimension_size
     }
-    fn strides(&self) -> Self {
-        Dim::new([1])
+
+    #[inline]
+    fn strides(&self) -> Self::IndexType {
+        [1]
     }
-    fn slice(&self) -> &[Ix] {
+
+    #[inline]
+    fn slice(&self) -> &[Idx] {
         &self.dimension_size
     }
-    fn stride_offset(index: &Self, strides: &Self) -> usize {
-        index[0] * strides[0]
-    }
+
+    #[inline]
     fn size_of_x_axis(&self) -> usize {
         self.dimension_size[0]
     }
+
+    #[inline]
     fn size_of_y_axis(&self) -> usize {
         1
     }
-    fn index_inside_dimension(index: &Self, dimension: &Self) -> bool {
-        index[0] < dimension[0]
-    }
-}
 
-impl Index<usize> for Dim1D {
-    type Output = usize;
-    fn index(&self, index: usize) -> &Self::Output {
-        &self.dimension_size()[index]
+    #[inline]
+    fn index_inside_dimension(&self, index: &Self::IndexType) -> bool {
+        index[0] < self.dimension_size[0]
     }
-}
 
-impl IndexMut<usize> for Dim1D {
-    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
-        &mut self.dimension_size_mut()[index]
+    #[inline]
+    fn linear_space_index_unchecked(&self, index: &Self::IndexType) -> usize {
+        let strides = self.strides();
+        index[0] * strides[0]
     }
-}
-
-impl Add for Dim1D {
-    type Output = Self;
 
-    fn add(self, rhs: Self) -> Self::Output {
-        let [x] = self.dimension_size;
-        let [other_x] = rhs.dimension_size;
-        Dim::new([x + other_x])
+    fn intersection(&self, other: &Self) -> Self {
+        let [self_x] = self.size_as_index();
+        let [other_x] = other.size_as_index();
+        Dim1D::new([min(self_x, other_x)])
     }
 }
 
-impl Sub for Dim1D {
-    type Output = Self;
-
-    fn sub(self, rhs: Self) -> Self::Output {
-        let [x] = self.dimension_size;
-        let [other_x] = rhs.dimension_size;
-        Dim::new([x - other_x])
+impl From<[usize; 1]> for Dim1D {
+    #[inline]
+    fn from(dimension_size: [usize; 1]) -> Self {
+        Dim1D::new(dimension_size)
     }
 }
 
-impl Mul<Ix> for Dim1D {
-    type Output = Self;
-
-    fn mul(self, rhs: Ix) -> Self::Output {
-        let [x] = self.dimension_size;
-        Dim::new([x * rhs])
+impl Index<usize> for Dim1D {
+    type Output = usize;
+    #[inline]
+    fn index(&self, index: usize) -> &Self::Output {
+        &self.dimension_size()[index]
     }
 }
 
-impl Div<Ix> for Dim1D {
-    type Output = Self;
-
-    fn div(self, rhs: Ix) -> Self::Output {
-        let [x] = self.dimension_size;
-        Dim::new([x / rhs])
+impl IndexMut<usize> for Dim1D {
+    #[inline]
+    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
+        &mut self.dimension_size_mut()[index]
     }
 }
 
 impl GridDimension for Dim2D {
-    type IndexPattern = Ix2;
-    const NDIM: usize = 2;
+    type IndexType = GridIdx2D;
+    // const NDIM: usize = 2;
+    #[inline]
     fn number_of_dimensions(&self) -> usize {
```

Comment:
Why did you remove the NDIM? Isn't the method inherently static or just constant?

Reply:
i can't impl the trait for a trait when there is a constant.

Reply:
Is this different to this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=4b1a2b4fc34623afa02ae34d827c2834

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3cf50b02133a0b8ca6badc66c9fb6190 but i guess it works if you really need it

Reply:
the NDIM thing is back and constants are used where possible..